File: conformance/code/individual-go/internal/service/enrichment_service.go
package service

import (
	"context"

	"individual/internal/clients"
	"individual/internal/common"
	"individual/internal/config"
	"individual/internal/models"

	"github.com/rs/zerolog/log"
)

// EnrichmentService handles data enrichment
type EnrichmentService interface {
	EnrichForCreate(ctx context.Context, individual *models.Individual, reqContext *models.RequestContext) error
	EnrichForUpdate(ctx context.Context, individual *models.Individual, reqContext *models.RequestContext) error
	EnrichForDelete(ctx context.Context, individual *models.Individual, reqContext *models.RequestContext) error
}

type enrichmentService struct {
	idgenClient clients.IDGenClient
	config      *config.IDGenConfig
}

// NewEnrichmentService creates a new enrichment service
func NewEnrichmentService(idgenClient clients.IDGenClient, config *config.IDGenConfig) EnrichmentService {
	return &enrichmentService{
		idgenClient: idgenClient,
		config:      config,
	}
}

// EnrichForCreate enriches individual for create operation
func (s *enrichmentService) EnrichForCreate(ctx context.Context, individual *models.Individual, reqContext *models.RequestContext) error {
	now := common.GetCurrentTimestamp()

	// Always generate a new UUID — id is server-managed and must not be
	// supplied by the client.
	individual.ID = common.GenerateUUID()

	// Always generate individualId via IDGen — readOnly per spec.
	customVars := map[string]string{"ORG": individual.TenantID}
	ids, err := s.idgenClient.GenerateIDs(ctx, individual.TenantID, s.config.Format, 1, customVars)
	if err != nil {
		log.Error().Err(err).Ctx(ctx).Str("tenantID", individual.TenantID).Msg("IDGen failed to generate individualId")
		// idgen is a downstream dependency — surface as DOWNSTREAM_ERROR (502) with the
		// specific cause, not the generic DATABASE_ERROR/500 catch-all.
		return common.ErrDownstream.WithContext(map[string]interface{}{
			"message": "failed to generate individualId: " + err.Error(),
		})
	}
	if len(ids) > 0 {
		individual.IndividualID = ids[0]
	} else {
		individual.IndividualID = "IND-" + common.GenerateUUID()[:8]
	}

	// Set audit fields (using clientID from header)
	if reqContext != nil {
		individual.CreatedBy = reqContext.UserID
		individual.ModifiedBy = reqContext.UserID
		individual.RequestID = reqContext.RequestID
	}
	individual.CreatedTime = now
	individual.ModifiedTime = now
	individual.RowVersion = 1
	// Newly created individuals should be active by default
	individual.Active = true

	// Enrich addresses
	for i := range individual.Addresses {
		if individual.Addresses[i].ID == "" {
			individual.Addresses[i].ID = common.GenerateUUID()
		}
		individual.Addresses[i].TenantID = individual.TenantID
		if reqContext != nil {
			individual.Addresses[i].CreatedBy = reqContext.UserID
			individual.Addresses[i].ModifiedBy = reqContext.UserID
			individual.Addresses[i].RequestID = reqContext.RequestID
		}
		individual.Addresses[i].CreatedTime = now
		individual.Addresses[i].ModifiedTime = now
	}

	// Enrich identifiers
	for i := range individual.Identifiers {
		if individual.Identifiers[i].ID == "" {
			individual.Identifiers[i].ID = common.GenerateUUID()
		}
		if reqContext != nil {
			individual.Identifiers[i].CreatedBy = reqContext.UserID
			individual.Identifiers[i].ModifiedBy = reqContext.UserID
			individual.Identifiers[i].RequestID = reqContext.RequestID
		}
		individual.Identifiers[i].CreatedTime = now
		individual.Identifiers[i].ModifiedTime = now
	}

	// NOTE: We do NOT auto-create a SYSTEM_GENERATED identifier when the
	// caller sends none. SYSTEM_GENERATED is just one of the allowed enum
	// values in identifierType — there is no spec business rule that mandates
	// auto-creation, and individualId (the IDGen-generated IND-XXXXX) already
	// covers the "stable cross-reference id" use case. See bug.md #1.

	// Enrich documents
	for i := range individual.Documents {
		if individual.Documents[i].ID == "" {
			individual.Documents[i].ID = common.GenerateUUID()
		}
		individual.Documents[i].IndividualID = individual.ID
		if reqContext != nil {
			individual.Documents[i].CreatedBy = reqContext.UserID
			individual.Documents[i].ModifiedBy = reqContext.UserID
			individual.Documents[i].RequestID = reqContext.RequestID
		}
		individual.Documents[i].CreatedTime = now
		individual.Documents[i].ModifiedTime = now
	}

	return nil
}

// EnrichForUpdate enriches individual for update operation
func (s *enrichmentService) EnrichForUpdate(ctx context.Context, individual *models.Individual, reqContext *models.RequestContext) error {
	now := common.GetCurrentTimestamp()

	// Update audit fields (using clientID from header)
	if reqContext != nil {
		individual.ModifiedBy = reqContext.UserID
		individual.RequestID = reqContext.RequestID
	}
	individual.ModifiedTime = now

	// Increment row version for optimistic locking
	individual.RowVersion++

	// Enrich addresses (present in the request => active under PUT full-replace)
	for i := range individual.Addresses {
		individual.Addresses[i].Active = true
		if individual.Addresses[i].ID == "" {
			individual.Addresses[i].ID = common.GenerateUUID()
			individual.Addresses[i].TenantID = individual.TenantID
			if reqContext != nil {
				individual.Addresses[i].CreatedBy = reqContext.UserID
				individual.Addresses[i].ModifiedBy = reqContext.UserID
				individual.Addresses[i].RequestID = reqContext.RequestID
			}
			individual.Addresses[i].CreatedTime = now
			individual.Addresses[i].ModifiedTime = now
		} else {
			if reqContext != nil {
				individual.Addresses[i].ModifiedBy = reqContext.UserID
				individual.Addresses[i].RequestID = reqContext.RequestID
			}
			individual.Addresses[i].ModifiedTime = now
		}
	}

	// Enrich identifiers (present in the request => active under PUT full-replace)
	for i := range individual.Identifiers {
		individual.Identifiers[i].Active = true
		if individual.Identifiers[i].ID == "" {
			individual.Identifiers[i].ID = common.GenerateUUID()
			if reqContext != nil {
				individual.Identifiers[i].CreatedBy = reqContext.UserID
				individual.Identifiers[i].ModifiedBy = reqContext.UserID
				individual.Identifiers[i].RequestID = reqContext.RequestID
			}
			individual.Identifiers[i].CreatedTime = now
			individual.Identifiers[i].ModifiedTime = now
		} else {
			if reqContext != nil {
				individual.Identifiers[i].ModifiedBy = reqContext.UserID
				individual.Identifiers[i].RequestID = reqContext.RequestID
			}
			individual.Identifiers[i].ModifiedTime = now
		}
	}

	// Enrich documents (present in the request => active under PUT full-replace)
	for i := range individual.Documents {
		individual.Documents[i].Active = true
		individual.Documents[i].IndividualID = individual.ID
		if individual.Documents[i].ID == "" {
			individual.Documents[i].ID = common.GenerateUUID()
			if reqContext != nil {
				individual.Documents[i].CreatedBy = reqContext.UserID
				individual.Documents[i].ModifiedBy = reqContext.UserID
				individual.Documents[i].RequestID = reqContext.RequestID
			}
			individual.Documents[i].CreatedTime = now
			individual.Documents[i].ModifiedTime = now
		} else {
			if reqContext != nil {
				individual.Documents[i].ModifiedBy = reqContext.UserID
				individual.Documents[i].RequestID = reqContext.RequestID
			}
			individual.Documents[i].ModifiedTime = now
		}
	}

	return nil
}

// EnrichForDelete enriches individual for delete operation
func (s *enrichmentService) EnrichForDelete(ctx context.Context, individual *models.Individual, reqContext *models.RequestContext) error {
	now := common.GetCurrentTimestamp()

	// Deactivate individual
	individual.Active = false

	// Update audit fields (using clientID from header)
	if reqContext != nil {
		individual.ModifiedBy = reqContext.UserID
	}
	individual.ModifiedTime = now

	// Deactivate identifiers
	for i := range individual.Identifiers {
		individual.Identifiers[i].Active = false
		individual.Identifiers[i].ModifiedTime = now
		if reqContext != nil {
			individual.Identifiers[i].ModifiedBy = reqContext.UserID
		}
	}

	return nil
}

File: conformance/code/individual/src/main/java/com/digit/individual/service/EnrichmentService.java
package com.digit.individual.service;

import com.digit.individual.client.IdgenClient;
import com.digit.individual.config.IndividualProperties;
import com.digit.individual.constants.ErrorCodes;
import com.digit.individual.model.Address;
import com.digit.individual.model.Document;
import com.digit.individual.model.Identifier;
import com.digit.individual.model.Individual;
import com.digit.individual.model.RequestContext;
import org.digit.tracer.model.CustomException;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

/**
 * Server-side enrichment for create/update/delete. Sets ids, audit fields, version, individualId
 * (via IDGen), and propagates them to nested entities.
 */
@Service
public class EnrichmentService {

    private final IdgenClient idgenClient;
    private final IndividualProperties.Idgen config;

    public EnrichmentService(IdgenClient idgenClient, IndividualProperties props) {
        this.idgenClient = idgenClient;
        this.config = props.getIdgen();
    }

    private static long now() {
        return System.currentTimeMillis();
    }

    private static String uuid() {
        return UUID.randomUUID().toString();
    }

    public void enrichForCreate(Individual ind, RequestContext reqContext) {
        long now = now();

        // id is server-managed.
        ind.setId(uuid());

        // individualId via IDGen (readOnly per spec).
        Map<String, String> customVars = new HashMap<>();
        customVars.put("ORG", ind.getTenantId());
        List<String> ids;
        try {
            ids = idgenClient.generateIds(ind.getTenantId(), config.getFormat(), 1, customVars);
        } catch (RuntimeException e) {
            // idgen is a downstream dependency — surface as DOWNSTREAM_ERROR (502) with the specific
            // cause, not the tracer's generic 500 catch-all. Mirrors Go enrichment_service.
            throw new CustomException(ErrorCodes.DOWNSTREAM,
                    "failed to generate individualId: " + e.getMessage(), HttpStatus.BAD_GATEWAY);
        }
        if (!ids.isEmpty()) {
            ind.setIndividualId(ids.get(0));
        } else {
            ind.setIndividualId("IND-" + uuid().substring(0, 8));
        }

        if (reqContext != null) {
            ind.setCreatedBy(reqContext.getUserId());
            ind.setModifiedBy(reqContext.getUserId());
            ind.setRequestId(reqContext.getRequestId());
        }
        ind.setCreatedTime(now);
        ind.setModifiedTime(now);
        ind.setRowVersion(1);
        ind.setActive(true);

        for (Address a : ind.getAddresses()) {
            if (a.getId() == null || a.getId().isEmpty()) {
                a.setId(uuid());
            }
            a.setTenantId(ind.getTenantId());
            if (reqContext != null) {
                a.setCreatedBy(reqContext.getUserId());
                a.setModifiedBy(reqContext.getUserId());
                a.setRequestId(reqContext.getRequestId());
            }
            a.setCreatedTime(now);
            a.setModifiedTime(now);
            a.setActive(true); // present in the request => active on create
        }

        for (Identifier i : ind.getIdentifiers()) {
            if (i.getId() == null || i.getId().isEmpty()) {
                i.setId(uuid());
            }
            if (reqContext != null) {
                i.setCreatedBy(reqContext.getUserId());
                i.setModifiedBy(reqContext.getUserId());
                i.setRequestId(reqContext.getRequestId());
            }
            i.setCreatedTime(now);
            i.setModifiedTime(now);
            i.setActive(true); // present in the request => active on create
        }

        for (Document d : ind.getDocuments()) {
            if (d.getId() == null || d.getId().isEmpty()) {
                d.setId(uuid());
            }
            d.setIndividualId(ind.getId());
            if (reqContext != null) {
                d.setCreatedBy(reqContext.getUserId());
                d.setModifiedBy(reqContext.getUserId());
                d.setRequestId(reqContext.getRequestId());
            }
            d.setCreatedTime(now);
            d.setModifiedTime(now);
            d.setActive(true); // present in the request => active on create
        }
    }

    public void enrichForUpdate(Individual ind, RequestContext reqContext) {
        long now = now();

        if (reqContext != null) {
            ind.setModifiedBy(reqContext.getUserId());
            ind.setRequestId(reqContext.getRequestId());
        }
        ind.setModifiedTime(now);
        ind.setRowVersion(ind.getRowVersion() + 1);

        for (Address a : ind.getAddresses()) {
            a.setActive(true); // present in the request => active under PUT full-replace
            if (a.getId() == null || a.getId().isEmpty()) {
                a.setId(uuid());
                a.setTenantId(ind.getTenantId());
                if (reqContext != null) {
                    a.setCreatedBy(reqContext.getUserId());
                    a.setModifiedBy(reqContext.getUserId());
                    a.setRequestId(reqContext.getRequestId());
                }
                a.setCreatedTime(now);
                a.setModifiedTime(now);
            } else {
                if (reqContext != null) {
                    a.setModifiedBy(reqContext.getUserId());
                    a.setRequestId(reqContext.getRequestId());
                }
                a.setModifiedTime(now);
            }
        }

        for (Identifier i : ind.getIdentifiers()) {
            i.setActive(true); // present in the request => active under PUT full-replace
            if (i.getId() == null || i.getId().isEmpty()) {
                i.setId(uuid());
                if (reqContext != null) {
                    i.setCreatedBy(reqContext.getUserId());
                    i.setModifiedBy(reqContext.getUserId());
                    i.setRequestId(reqContext.getRequestId());
                }
                i.setCreatedTime(now);
                i.setModifiedTime(now);
            } else {
                if (reqContext != null) {
                    i.setModifiedBy(reqContext.getUserId());
                    i.setRequestId(reqContext.getRequestId());
                }
                i.setModifiedTime(now);
            }
        }

        for (Document d : ind.getDocuments()) {
            d.setActive(true); // present in the request => active under PUT full-replace
            d.setIndividualId(ind.getId());
            if (d.getId() == null || d.getId().isEmpty()) {
                d.setId(uuid());
                if (reqContext != null) {
                    d.setCreatedBy(reqContext.getUserId());
                    d.setModifiedBy(reqContext.getUserId());
                    d.setRequestId(reqContext.getRequestId());
                }
                d.setCreatedTime(now);
                d.setModifiedTime(now);
            } else {
                if (reqContext != null) {
                    d.setModifiedBy(reqContext.getUserId());
                    d.setRequestId(reqContext.getRequestId());
                }
                d.setModifiedTime(now);
            }
        }
    }

    public void enrichForDelete(Individual ind, RequestContext reqContext) {
        long now = now();
        ind.setActive(false);
        if (reqContext != null) {
            ind.setModifiedBy(reqContext.getUserId());
        }
        ind.setModifiedTime(now);
        for (Identifier i : ind.getIdentifiers()) {
            i.setActive(false);
            i.setModifiedTime(now);
            if (reqContext != null) {
                i.setModifiedBy(reqContext.getUserId());
            }
        }
    }
}
